### PR TITLE
Fixes #10806 - Introduce RequestLogWrapper to address SocketAddress concerns

### LINKED ISSUE #10806: Can't always access remote address from RequestLog
**Jetty version(s)**
12.0.2

**Jetty Environment**
ee10

**Java version/vendor** 
OpenJDK Runtime Environment Corretto-17.0.5.8.1 (build 17.0.5+8-LTS)

**OS type/version**
MacOS

**Description**
Using `CustomRequestLog` with the `NCSA_FORMAT`, the remote address is often (but not always) missing from the resulting logs. There doesn't seem to be any pattern to this, often two runs of the exact same test will produce different results (sometimes the remote address is present, other times it is blank).

In debugging through what's going on, I've pinpointed this down to:

* CustomRequestLog uses `Remote.getRemoteAddr(request)` here:

https://github.com/jetty/jetty.project/blob/7be6c4b375de7cf44f71f4f7e7f1c87ed386ef9b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/CustomRequestLog.java#L821-L824

* getRemoteAddr calls `getRemoteSocketAddress` here:

https://github.com/jetty/jetty.project/blob/7be6c4b375de7cf44f71f4f7e7f1c87ed386ef9b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java#L391-L395

* ...but `SocketChannelImpl` has a check that the socket is still open, but finds the socket closed by the time the RequestLog is called and therefore throws an exception:

https://github.com/openjdk/jdk/blob/1ec0d02717b6be4faeb13cd0596d80eea90e81ed/src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java#L255-L260

### PR DESCRIPTION
**Pull request by joakime**
Alternate approach from PR #10815 

Uses a `RequestLogWrapper` to address requestlog concerns, instead of modifying the low level IO layer.

Fixes #10806

### CODE REVIEW

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestLogTest.java [hunk 1]
```diff
@@ -359,40 +349,118 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
+                    response.getHeaders().put(HttpHeader.CONTENT_TYPE, "text/plain; charset=utf-8");
+
+                    String msg = "Got %s to %s%n".formatted(request.getMethod(), request.getHttpURI());
+                    Callback testCallback = Callback.from(callback, () ->
+                    {
+                        EndPoint endPoint = request.getConnectionMetaData().getConnection().getEndPoint();
+                        // Close connection
+                        endPoint.close();
+                        // Wait for endpoint to be closed
+                        Awaitility.await().atMost(Duration.ofSeconds(5)).until(() -> !endPoint.isOpen());
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/RequestLogTest.java:419]: Please `static import Awaitility.await` like we do with other testing facilities.

**Review by sbordet — APPROVED** [review]
LGTM beside the nit about Awaitility.
Much simpler solution than the other PR.

**Review by gregw — CHANGES_REQUESTED** [review]
I think I like the other approach more.  I don't like putting on another wrapper just to remember address data that might go away.

I prefer the approach that an end point has local/remote address regardless of it's open/close state.   In this case we have been caught by the request log, but I can imagine other situations (specially debugging) that we are racing with a close to get the addresses.

Far better to make the the addresses on an endpoint immutable rather than always wrap them in case they are not.

**Comment by gregw** [comment]
@sbordet @joakime I now agree that #10815 is wrong to fix the remote/local address in the EndPoint, as that is intrinsically dynamic.

This PR is correct in that the solution is to make the addresses immutable for the duration of a request cycle.    But it is doing that wrong,  as it is replacing a probably immutable ConnectionMetaData with a definitely immutable one.  So it is often wasted effort.

I think the correct solution is that ConnectionMetaData should always be immutable.  I.E any Connection that provides a ConnectionMetaData for a request cycle should provide a non-dynamic instance - so it will ask the endpoint at request init time what the addresses are.   That ConnectionMetaData will then never change.   The Connection can then decide if it reuses it for subsequent requests or creates a new one for every request.

**Comment by gregw** [comment]
Note that the ConnectionMetaData is passed to the constructor of the HttpChannel, so that any connection type that does support changing addresses for remote/local will need to recreate the HttpChannel instance.
So the change here is that h3 will need to remake the ConnectionMetaData and HttpChannel's whenever it detects an address has changed (or maybe on every request cycle, but that could be slow).

@sbordet are h2/h3 recycling HttpChannel instances at the moment?   

@lorban at some stage we need to evaluate if recycling HttpChannels is worth it.... we can probably test that easily in HTTP/1, which does reuse, by making a version that does not.

**Comment by gregw** [comment]
@joakime I've implemented my suggested approach, of making the addresses immutable within the `ConnectionMetaData`, which in practise for h1, h2 & FCGI is in the `Connection`, but for h3 is a new `ConnectionMetaData` in each `onRequest` call.

Note that this also revealed that only h2 was correctly respecting `HttpConfiguration.getLocalAddress()`.  I have now implemented for h1, h2, h3 & FCGI.

**Comment by joakime** [comment]
Dropping in favor of #10867

**Pull request closed without merging**